synapse/push/push_types.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, List, Optional
+
+from typing_extensions import TypedDict
+
+
+class EmailReason(TypedDict, total=False):
```

Comment:
The `total=False` isn't great in my opinion, but it's currently necessary as many code path iteratively construct these dicts. Maybe reworking that could be done as a next step?

Reply:
Can we keep them as normal dicts while we're building them, and then cast? Or just use local variables and then build it at the end? e.g.:

```python
let room_id = ...
...
let room_name = ...
....
reason: EmailReason  = { ... }
```

Not sure if that works

Reply:
That would require changing how these functions iteratively build the struct, while i think that's a good idea in general, i don't know how possible that is, certainly as i remember some of them building these up across multiple functions, passing intermediate results.

Reply:
Yup, I also believe that's the case. I'll take another look and see if I can resolve at least some easy cases here.

Reply:
I looked again and with the current code structure doing anything like this doesn't really make sense as for the reasons @ShadowJonathan mentioned.

The thing we could do is creating None value fields instead of structures with missing fields (so adding a bunch of Optionals instead of `total=False`). I currently have no opinion if this is better or worse.

Anything else would require more restructuring of the code, which I think is better done seperately?

Reply:
Thanks for looking into it. I think restructuring this may be the right answer eventually but you're right that it's not for here and now.